core/network/src/service.rs

```diff
@@ -185,63 +175,86 @@ impl<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT> Service<B, S,
 		self.network.lock().average_upload_per_sec()
 	}
 
+	/// Get a clone of the channel to network/libp2p.
+	pub fn network_sender(&self) -> NetworkChan {
+		self.network_sender.clone()
+	}
+
 	/// Called when a new block is imported by the client.
 	pub fn on_block_imported(&self, hash: B::Hash, header: &B::Header) {
-		self.handler.on_block_imported(&mut NetSyncIo::new(&self.network, self.protocol_id), hash, header)
+		let _ = self
+			.protocol_sender
+			.send(ProtocolMsg::BlockImported(hash, header.clone()));
 	}
 
 	/// Called when a new block is finalized by the client.
 	pub fn on_block_finalized(&self, hash: B::Hash, header: &B::Header) {
-		self.handler.on_block_finalized(&mut NetSyncIo::new(&self.network, self.protocol_id), hash, header)
+		let _ = self
+			.protocol_sender
+			.send(ProtocolMsg::BlockFinalized(hash, header.clone()));
 	}
 
 	/// Called when new transactons are imported by the client.
 	pub fn trigger_repropagate(&self) {
-		self.handler.propagate_extrinsics(&mut NetSyncIo::new(&self.network, self.protocol_id));
+		let _ = self.protocol_sender.send(ProtocolMsg::PropagateExtrinsics);
 	}
 
 	/// Make sure an important block is propagated to peers.
 	///
 	/// In chain-based consensus, we often need to make sure non-best forks are
 	/// at least temporarily synced.
 	pub fn announce_block(&self, hash: B::Hash) {
-		self.handler.announce_block(&mut NetSyncIo::new(&self.network, self.protocol_id), hash);
+		let _ = self.protocol_sender.send(ProtocolMsg::AnnounceBlock(hash));
 	}
 
 	/// Send a consensus message through the gossip
 	pub fn gossip_consensus_message(&self, topic: B::Hash, message: Vec<u8>, broadcast: bool) {
-		self.handler.gossip_consensus_message(
-			&mut NetSyncIo::new(&self.network, self.protocol_id),
-			topic,
-			message,
-			broadcast,
-		)
-	}
-	/// Execute a closure with the chain-specific network specialization.
-	pub fn with_spec<F, U>(&self, f: F) -> U
-		where F: FnOnce(&mut S, &mut Context<B>) -> U
-	{
-		self.handler.with_spec(&mut NetSyncIo::new(&self.network, self.protocol_id), f)
+		let _ = self
+			.protocol_sender
+			.send(ProtocolMsg::GossipConsensusMessage(
+				topic, message, broadcast,
+			));
 	}
 
 	/// access the underlying consensus gossip handler
-	pub fn consensus_gossip<'a>(&'a self) -> &'a RwLock<ConsensusGossip<B>> {
-		self.handler.consensus_gossip()
+	pub fn consensus_gossip_messages_for(
+		&self,
+		topic: B::Hash,
+	) -> mpsc::UnboundedReceiver<ConsensusMessage> {
+		let (sender, port) = unbounded();
+		let _ = self
+			.protocol_sender
+			.send(ProtocolMsg::GossipConsensusMessagesFor(topic, sender));
+		port.recv().unwrap()
```

Comment:
no unwraps allowed in production code.

Reply:
use `expect` with a proof in the form `"fact 1; fact 2; ...; fact n; qed"` where the facts prove that the thing can't panic. It is really not recommended to rely on any assumptions outside of the module.